src/Microsoft.OData.Core/Json/JsonWriterAsync.cs

```diff
@@ -0,0 +1,448 @@
+        }
+
+        /// <summary>
+        /// Write an sbyte value.
+        /// </summary>
+        /// <param name="value">SByte value to be written.</param>
+        public async Task WriteValueAsync(sbyte value)
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteValueAsync(value).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write a string value.
+        /// </summary>
+        /// <param name="value">String value to be written.</param>
+        public async Task WriteValueAsync(string value)
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteValueAsync(value, this.stringEscapeOption, this.wrappedBuffer)
+                .ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write a byte array.
+        /// </summary>
+        /// <param name="value">Byte array to be written.</param>
+        public async Task WriteValueAsync(byte[] value)
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteValueAsync(value, this.wrappedBuffer, ArrayPool).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write a raw value.
+        /// </summary>
+        /// <param name="rawValue">Raw value to be written.</param>
+        public async Task WriteRawValueAsync(string rawValue)
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteAsync(rawValue).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Clears all buffers for the current writer.
+        /// </summary>
+        public async Task FlushAsync()
+        {
+            await this.writer.FlushAsync().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Start the stream property valuescope.
+        /// </summary>
+        /// <returns>The stream to write the property value to</returns>
+        public async Task<Stream> StartStreamValueScopeAsync()
+        {
+            await this.WriteValueSeparatorAsync().ConfigureAwait(false);
+            await this.writer.WriteAsync(JsonConstants.QuoteCharacter).ConfigureAwait(false);
+            await this.writer.FlushAsync().ConfigureAwait(false);
+            // IMPORTANT: The Dispose method of the returned stream does the following:
+            // - Writes trailing bytes to the writer synchronously
+            // - Flushes the buffer of the writer synchronously
+            // ODL supports net45 and netstandard1.1 (in addition to .netstandard2.0)
+            // This makes it complicated to implement IAsyncDisposable in ODataBinaryStreamWriter
+            // TODO: Can the returned stream be safely used asynchronously?
+            this.binaryValueStream = new ODataBinaryStreamWriter(writer, this.wrappedBuffer, ArrayPool);
+            return this.binaryValueStream;

```

Comment:
If writing and flushing synchronously when disposing is a great concern, maybe you could extend the `ODataBinaryStreamWriter` with a version where `Dispose()` simply calls the `base.Dispose()` then you add a method like `WriteTrailingBytesAndFlushAsync()`. Then you call:
```c#
await (this.binaryValueStream as ODataBinaryStreamWriterAsync).WriteTrailingBytesAndFlushAsync();
this.binaryValueStream.Dispose();
this.binaryValueStream = null;
```

But this looks messy and maybe not worth it if it will be replaced when dropping net45.

Reply:
But since the `binaryValueStream` property is private implementation detail, maybe there's no harm in changing its type to some custom `AsyncDisposableStream` that extends `Stream` and adds a `DisposeAsync` method. Then make `ODataBinaryStreamWriterAsync` extend that type. Then when `net45` is dropped, you could extend a regular `Stream`.

Or what about an extension method to the `Stream` class wrapped instead `#if NET45`?